### ISSUE TITLE
ts-node-dev sometimes uses old versions of files for compilation

### ISSUE THREAD

**Issue by arbieo** [whitecolor/ts-node-dev#57]
I am seeing ts-node-dev fail to hot reload files sometimes (even though it notices the file changes and tries to restart according to the output). I don't have any steps to repro and I'm seeing this randomly every 2-3 days.

It seems different enough from the other issues i've seen on here to make its own thread.

I will fully explain my workflow for the most recent case as it may shed some light.

the error looked like:

[INFO] 15:58:29 Restarting: /Users/ryan/app_engine/StoryDoodle/src/server/util/redis.ts has been modified
Using ts-node version 8.0.2, typescript version 3.3.3333
[ERROR] 15:58:32 ⨯ Unable to compile TypeScript:
src/server/cache/JSONCache.ts(5,34): error TS2551: Property 'getAsync' does not exist on type 'IPromisifedRedis'. Did you mean 'hmgetAsync'?
src/server/cache/JSONCache.ts(16,17): error TS2339: Property 'setAsync' does not exist on type 'IPromisifedRedis'.

This is a normal and explainable error but it was only possible with an old file of /util/redis.ts because that's the only time I had hmgetAsync type defined. In addition, other files which depend on this file are compiling correctly. (they should fail on that version of the file)

My flow:

1. Make /util/redis.ts with version A (which would not have this error)
2. Make another file that debends on and works with /util/redis.ts
3. Modify /util/redis.ts to version B (which should produce this error)
4. Change my mind and change /util/redis.ts back to version A
5. Compilation is fine.
6. Create a new file that depends on /util/redis.ts
7. Get this error (which could only occur on version B of /util/redis.ts)
8. Try to resolve error by editing files. This causes visible reloads but error still persists
9. Restarting ts-node-dev compiles properly

my command: ts-node-dev --respawn src/server/index.ts
OS: MAC OSX 10.14
ts-node-dev version: 1.0.0-pre.32

I can do additional debugging the next time this happens. Just let me know what might help.

**Comment by arbieo** [comment]
After having other things failing, I suspect this may be a silently caught error somewhere related to OSX hitting the max number of files. If you are seeing similar strange behaviour in OSX I would sugget running the following (stolen from elsewhere on the internet):

`$ echo kern.maxfiles=65536 | sudo tee -a /etc/sysctl.conf
$ echo kern.maxfilesperproc=65536 | sudo tee -a /etc/sysctl.conf
$ sudo sysctl -w kern.maxfiles=65536
$ sudo sysctl -w kern.maxfilesperproc=65536
$ ulimit -n 65536`

**Comment by whitecolor** [comment]
Maybe something like this possible, if you find a way to reproduce it consistently, please show it.

**Comment by PachowStudios** [comment]
Also experiencing this issue. My reproduction steps are basically the same as the issue author.

Here's the simplified version:
1. Create file util.ts with a function accepting a string.
2. Create file test.ts which calls the function in util.ts, but passes in a number instead of a string (which rightfully causes a compilation error).
3. Modify the function in util.ts to accept a number, which should fix the compile error.
4. The compile error persists unless test.ts is modified to have different content. Just saving the file again will cause a recompile, but the error still persists.
5. Restarting ts-node-dev completely fixes the error.

This is extremely annoying to deal with and results in having to restart the compilation quite often, which can take close to a minute.

**Comment by whitecolor** [comment]
@PachowStudios maybe you could create a repo to reproduce?

**Comment by saiya** [comment]
I think this issue is same as #60 . As I commented in #60, there is a non-intuitive (but NOT bug) behavior due to combination of TypeScript nature and incremental compilation.

**Comment by Akryum** [comment]
I think it should considered as a bug since it prevents the app from starting without killing and restarting the whole process.

**Comment by rhyek** [comment]
Yes, I totally agree which is why I stopped using `ts-node-dev` and opted for a combination of `tsc-watch` (for the `first_succes` and `success` options it offers) and `node-dev` and it's been perfect. No issues whatsoever, but the config is not as straightforward. There are a few gotchas. Mainly, `--noEmitOnErrors` does not work with tsc's `--watch` option (which `tsc-watch` uses internally).

**Comment by Akryum** [comment]
Or webpack :D

**Comment by whitecolor** [comment]
> The compile error persists unless test.ts is modified to have different content. Just saving the file again will cause a recompile, but the error still persists.

I'll try to fix this will reinitializing compilation on compiler's error, not sure how it will affect large codebase, but this is the only way to get rid of this.

**Comment by PachowStudios** [comment]
Maybe a potential solution would be to just recompile files with errors (after dumping the cache) instead of restarting the entire compilation?

Restarting the compilation will still likely result in this package being unusable.

**Comment by seiyria** [comment]
I've been running into this too, except fairly frequently (on windows). It seems like every time I change a function signature or add a new file, I need to restart the process or it throws errors.

**Comment by plaa** [comment]
I've just started having similar issues constantly, where the Node server is restarted but uses old version of code, without any errors being present.  Killing and restarting ts-node-dev takes the new code into use.  This didn't happen earlier, I'm unsure what has changed.  I think I upgraded Node in between.

I'm able to reproduce the problem with the following steps:

1. Start the server with ts-node-dev
2. Call an API with `curl` (not strictly necessary, reproduces without this step)
3. Modify the controller to add `console.log('HERE')`
4. Save the file --> the server restarts
5. Call the API with `curl`  -->  **'HERE' does not get printed**
6. Modify the controller line to `console.log('HERE2')`
7. Save the file --> the server restarts
8. Call the API with `curl` -->  **'HERE2' is printed**

I've tested this on 1.0.0-pre.50 (I'm unable to test on newer versions due to #208).  Running on macOS.  Raising the `ulimit -n 30000` limit did not help, the suggested `sysctl` values are already >40000.

Any ideas?

**Comment by whitecolor** [comment]
@plaa make reproducible example repository.

**Comment by plaa** [comment]
@whitecolor I minimized the repo and finally figured out that if I remove the explicit `--watch 'src'` option, it starts reloading the files properly.  This seems like a clear bug to me.

Example repo with reproduction instructions: https://github.com/plaa/ts-node-dev-watch-bug

**Comment by whitecolor** [comment]
I'll take a look.

**Comment by whitecolor** [comment]
@plaa  why you are watching the required code that is already watched? This definitely may lead to some problems. `--watch` option is for watching arbitrary files/files, it is not intended to be used this way.

**Comment by plaa** [comment]
@whitecolor Configuration was like this when I joined the project.  Possibly because there's some configuration / YAML files under `src/` to be watched as well.

In any case, I wouldn't expect specifying files to be watched multiple times to cause such erratic behaviour. It should either cause a clear error, or (preferably) just ignore duplicates.

**Comment by whitecolor** [comment]
@plaa ok will see how to resolve it.

**Comment by craigphicks** [comment]
I'm having the same same problem with `ts-node`, version `v9.1.1`.  The file under test was renamed from 'test.ts' to 'testset1.ts'.  Compiling with `tsc` and executing the js file runs correctly.  But `ts-node` seems to be using a cached old version -- perhaps the new name has messed up the method by which it detects a changed source file?  

The `npm` page for `ts-node` say there is a flag `--no-cache`.  However that is not recognized by the version I am using.  Attempts to find the cache did not succeed.  If @wclr could inform a workaround method for deleting the cache that would be helpful.

**Comment by LvChengbin** [comment]
any update on this issue?

**Comment by patostickar** [comment]
I'm having a different but similar issue. I have a .ts file used as a mock DB where I added more info. I have no compilation errors after modifying it and updating the types of the request response, but still, I'm being served the old file info when making the GET request. I console.logged the request response, and you can see that it's different if I run ts-node-dev vs running the build. Specifically, the entries array is not being displayed.

![dev](https://user-images.githubusercontent.com/58763654/196798765-f588eea5-6281-489b-9b37-4df62fa567ac.jpg)

![build](https://user-images.githubusercontent.com/58763654/196798780-487487b0-465e-4268-9950-acb43d654d07.jpg)

**Comment by gaboesquivel** [comment]
+ 1
